velox/exec/HashBuild.cpp

```diff
@@ -199,6 +214,53 @@ RowTypePtr HashBuild::inputType() const {
                             : joinNode_->sources()[1]->outputType();
 }
 
+void HashBuild::setupFilter() {
+  ExprSet exprs({joinNode_->filter()}, operatorCtx_->execCtx());
+  VELOX_DCHECK_EQ(exprs.exprs().size(), 1);
+  auto& expr = exprs.expr(0);
+  filterPropagatesNulls_ = expr->propagatesNulls();
+  if (filterPropagatesNulls_) {
+    auto outputType = joinNode_->sources()[1]->outputType();
+    for (auto& field : expr->distinctFields()) {
+      if (auto index = outputType->getChildIdxIfExists(field->field());
+          index.has_value()) {
+        filterChannels_.push_back(*index);
+      }
+    }
+  }
+}
+
+void HashBuild::removeFilterInputNullRows() {
+  bool changed = false;
+  VELOX_DCHECK(
+      std::is_sorted(dependentChannels_.begin(), dependentChannels_.end()));
+  for (auto i : filterChannels_) {
```

Comment:
Would it make sense to initialize 2 lists in the setupFilter: (1) a list of indices into table_->hashers() for filter columns; (2) a list of indices into decoders_ for filter columns. This way we won't need to perform additional lookups for each batch.